_posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd

```diff
@@ -0,0 +1,283 @@
+
+[^2]: @RonnebergerFB15
+
+1.  It demonstrates how to perform data augmentation for an image segmentation task.
+
+2.  It uses [luz](https://mlverse.github.io/luz/), `torch`'s high-level interface, to train the model.
+
+3.  It [JIT-traces](https://blogs.rstudio.com/ai/posts/2021-08-10-jit-trace-module/) the trained model and saves it for deployment on mobile devices. (JIT being the acronym commonly used for the `torch` just-in-time compiler.)
+
+4.  It includes proof-of-concept code (though not a discussion) of the saved model being run on Android.
+
+And if you think that this in itself is not exciting enough -- our task here is to find cats and dogs. What could be more helpful than a mobile application making sure you can distinguish your cat from the fluffy sofa she's reposing on?
+
+![A cat from the Oxford Pet Dataset (@parkhi12a).](images/Abyssinian_47.jpg)
+
+## Train in R
+
+We start by preparing the data.
+
+### Pre-processing and data augmentation
+
+As provided by `torchdatasets`, the [Oxford Pet Dataset](https://www.robots.ox.ac.uk/~vgg/data/pets/) comes with three variants of target data to choose from: the overall class (cat or dog), the individual breed (there are thirty-seven of them), and a pixel-level segmentation with three categories: foreground, boundary, and background. The latter is the default; and it's exactly the type of target we need.
+
+A call to `oxford_pet_dataset(root = dir)` will trigger the initial download:
+
+```{r}
+# need torch > 0.6.1
+# may have to run remotes::install_github("mlverse/torch") depending on when you read this
+library(torch) 
+library(torchvision)
+library(torchdatasets)
+library(luz)
+
+dir <- "~/.torch-datasets/oxford_pet_dataset"
+
+ds <- oxford_pet_dataset(root = dir)
+```
+
+Images (and corresponding masks) come in different sizes. For training, however, we'll need all of them to be the same size. This can be accomplished by passing in `transform =` and `target_transform =` arguments. But what about data augmentation (basically always a useful measure to take)? Imagine we make use of random flipping. An input image will be flipped -- or not -- according to some probability. But if the image is flipped, the mask better had be, as well! Input and target transformations are not independent, in this case.
+
+A solution is to create a wrapper around `oxford_pet_dataset()` that lets us "hook into" the `.getitem()` method, like so:
+
+```{r}
+pet_dataset <- torch::dataset(
+  
+  inherit = oxford_pet_dataset,
+  
+  initialize = function(..., size, normalize = TRUE, augmentation = NULL) {
+    
+    self$augmentation <- augmentation
+    
+    input_transform <- function(x) {
+      x <- x %>%
+        transform_to_tensor() %>%
+        transform_resize(size) 
+      if (isTRUE(normalize)) x <- x %>%
+        transform_normalize(mean = c(0.485, 0.456, 0.406),
```

Comment:
Maybe add a quick comment on why we use these numbers on normalize?

Reply:
will do!